github.com/cockroachdb/pebble/internal/metamorphic: TestMeta failed

github.com/cockroachdb/pebble/internal/metamorphic.TestMeta [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicRace/19730607?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicRace/19730607?buildTab=artifacts#meta) on refs/heads/master @ [886faf44959c](https://github.com/cockroachdb/pebble/commits/886faf44959c):


```
        db1.LogData("buqf") // <nil> #1283
        db1.Get("tzpviiukfkvj@8") // [""] pebble: not found #1285
        iter10.Last() // [true,"qlrcfqep",<no point>,["qlrcfqep","tzpviiukfkvj")=>{"@14"="tis"}*] <nil> #1257
        // 04:08:05.690 INFO: [JOB 61] flushing: sstable created 000204
        iter11.SeekGE("ebrymcre@3", "jnxxumbnom") // [invalid] <nil> #1024
        // 04:08:05.690 INFO: [JOB 61] flushing: blob file created 000205
        iter14.Last() // [true,"vsmwpphq@4","wkastazzikvfcqrhjc",["qlrcfqep","vsmwpphq@3")=>{"@16"="mnoswcnzvdr","@14"="tis"}] <nil> #1159
        iter11.Last() // [true,"rhnechgslbl@8","dqydguc",["qlrcfqep","trqqlr")=>{"@14"="tis"}*] <nil> #1033
        db1.Get("ebrymcre") // ["r"] <nil> #1288
        iter10.SeekLT("lvpzouaq@13", "") // [true,"jnuevrlsej",<no point>,["jnuevrlsej","nzpqm")=>{"@12"="xjknxmia"}*] <nil> #1258
        iter14.SeekGE("vsmwpphq@3", "") // [false] <nil> #1171
        iter11.Last() // [true,"rhnechgslbl@8","dqydguc",["qlrcfqep","trqqlr")=>{"@14"="tis"}] <nil> #1042
        iter10.Last() // [true,"qlrcfqep",<no point>,["qlrcfqep","tzpviiukfkvj")=>{"@14"="tis"}*] <nil> #1259
        iter10.SeekPrefixGE("dvvvji") // [false] <nil> #1260
        iter10.SeekGE("zuqmrgzcoda@6", "") // [false] <nil> #1265
        iter10.NextPrefix() // [false] <nil> #1300
        iter15.Last() // [true,"zuqmrgzcoda@2","uesputayzibylqmhsilx",<no range>*] <nil> #1196
        iter11.SeekPrefixGE("rhnechgslbl") // [true,"rhnechgslbl",<no point>,["rhnechgslbl","rhnechgslbl\x00")=>{"@14"="tis"}*] <nil> #1044
        iter11.InternalNext() // <nil> #1046
        iter11.Prev("") // [false] pebble: unsupported reverse prefix iteration #1049
        iter11.SetBounds("cisrojzhbowt@15", "gdmkzfkjho@5") // <nil> #1050
        iter12.SeekLT("zayoouirqgnl@4", "") // [true,"cisrojzhbowt@10","hadbpmrxsmqri",["cisrojzhbowt","cisrojzhbowt@5")=>{"@4"="uqhzp"}*] <nil> #1163
        iter11.SeekLT("gdmkzfkjho@5", "") // [true,"gdmkzfkjho@7","diaebrhzwl",<no range>] <nil> #1051
        iter13.SeekLT("zntcdtawywu", "") // [true,"zayoouirqgnl@9","qywfuhikebiwx",<no range>] <nil> #1510
        iter15.Next("") // [false] <nil> #1206
        iter10.First() // [true,"cisrojzhbowt",<no point>,["cisrojzhbowt","dvvvji")=>{"@4"="uqhzp"}*] <nil> #1305
        iter10.Next("") // [true,"jnuevrlsej",<no point>,["jnuevrlsej","nzpqm")=>{"@12"="xjknxmia"}*] <nil> #1327
        db1.Get("jnxxumbnom@11") // ["wpygyzixypbvzjovrzd"] <nil> #1289
        db1.Get("zuqmrgzcoda@15") // [""] pebble: not found #1290
        db1.Set("vsmwpphq@9", "tiemmprjudjdmu") // <nil> #1294
        // 04:08:05.699 INFO: [JOB 61] flushing: sstable created 000206
        iter13.Next("") // [false] <nil> #1511
        // 04:08:05.700 INFO: [JOB 61] flushing: blob file created 000207
        iter12.SeekGE("vdwiz@13", "") // [false] <nil> #1170
        iter15.First() // [true,"cisrojzhbowt","lmzx",["cisrojzhbowt","dvvvji")=>{"@4"="uqhzp"}*] <nil> #1209
        iter15.Prev("") // [false] <nil> #1212
        iter10.SeekLT("jnxxumbnom@1", "") // [true,"jnuevrlsej",<no point>,["jnuevrlsej","nzpqm")=>{"@12"="xjknxmia"}] <nil> #1328
        db1.Set("vsmwpphq@4", "movgikl") // <nil> #1295
        // 04:08:05.703 INFO: [JOB 61] flushing: sstable created 000208
        iter11.Next("") // [false] <nil> #1052
        // 04:08:05.704 INFO: [JOB 61] flushing: blob file created 000209
        iter14.SeekLT("gdmkzfkjho@1", "") // [false] <nil> #1186
        iter14.Prev("") // [false] <nil> #1198
        db1.Get("vsmwpphq@4") // ["movgikl"] <nil> #1298
        iter13.Last() // [true,"zayoouirqgnl@9","qywfuhikebiwx",<no range>] <nil> #1519
        db1.Set("jnxxumbnom@1", "rjy") // <nil> #1304
        iter12.First() // [true,"cisrojzhbowt","lmzx",["cisrojzhbowt","cisrojzhbowt@5")=>{"@4"="uqhzp"}*] <nil> #1176
        
        To reduce:  go test ./internal/metamorphic -tags invariants -run 'TestMeta$' --run-dir /artifacts/meta/250520-040804.1933748673195/random-023 --try-to-reduce -v
--- FAIL: TestMeta/execution/random-023 (1.57s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run 'TestMeta$' ./internal/metamorphic -seed 1747714084193886555 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMeta.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: PEBBLE-448

--- a/data_test.go
+++ b/data_test.go
@@ -136,6 +136,8 @@ func runIterCmd(d *datadriven.TestData, iter *Iterator, closeIter bool) string {
 					op = "seeklt"
 				}
 				fmt.Fprintf(&b, "%s=%q\n", field, op)
+			case "value.Len()":
+				fmt.Fprintf(&b, "%s=%d\n", field, iter.value.Len())
 			default:
 				return fmt.Sprintf("unrecognized inspect field %q\n", field)
 			}

--- a/iterator.go
+++ b/iterator.go
@@ -1002,6 +1002,10 @@ func (i *Iterator) findPrevEntry(limit []byte) {
 			// must've already iterated over it.
 			// This is the final entry at this user key, so we may return
 			i.rangeKey.rangeKeyOnly = i.iterValidityState != IterValid
+			if i.rangeKey.rangeKeyOnly {
+				// The point iterator is now invalid, so clear the point value.
+				i.value = base.InternalValue{}
+			}
 			i.keyBuf = append(i.keyBuf[:0], key.UserKey...)
 			i.key = i.keyBuf
 			i.iterValidityState = IterValid

--- a/testdata/iter_histories/blob_references
+++ b/testdata/iter_histories/blob_references
@@ -195,3 +195,30 @@ l: (peach, .)
 stats: seeked 1 times (1 internal); stepped 11 times (11 internal); blocks: 0B cached, 341B not cached (read time: 0s); points: 12 (12B keys, 24B values); separated: 12 (96B, 96B fetched)
 .
 stats: seeked 1 times (1 internal); stepped 12 times (12 internal); blocks: 0B cached, 341B not cached (read time: 0s); points: 12 (12B keys, 24B values); separated: 12 (96B, 96B fetched)
+
+# Regression test for #4741.
+#
+# Previously an iterator traversing in reverse direction would not reset i.value
+# to an empty InternalValue if it landed on a range key. While i.Value() would
+# never use the garbage value in i.value, maybeSampleRead would in order to
+# retrieve the value length. This resulted in a data race if the underlying
+# sstable iterator backing a LazyValue stored in i.value was pooled and reused.
+
+define verbose format-major-version=21
+L6
+  rangekey:a-d:{(#5,RANGEKEYSET,@2,foo)}
+  b.SETWITHDEL.2:blob{fileNum=000009 value=keylime}
+----
+L6:
+  000004:[a#5,RANGEKEYSET-d#inf,RANGEKEYSET] seqnums:[2-5] points:[b#2,SETWITHDEL-b#2,SETWITHDEL] ranges:[a#5,RANGEKEYSET-d#inf,RANGEKEYSET] size:1029 blobrefs:[(000009: 7); depth:1]
+Blob files:
+  000009: 63 physical bytes, 7 value bytes
+
+combined-iter
+seek-lt c
+seek-lt b
+inspect value.Len()
+----
+b: (keylime, [a-d) @2=foo UPDATED)
+a: (., [a-d) @2=foo)
+value.Len()=0

